tests/tracer/test_propagation.py

```diff
@@ -76,3 +79,35 @@ def test_WSGI_extract(self):
 class TestPropagationUtils(object):
     def test_get_wsgi_header(self):
         assert get_wsgi_header("x-datadog-trace-id") == "HTTP_X_DATADOG_TRACE_ID"
+
+    def test_from_wsgi_header(self):
+        assert from_wsgi_header("HTTP_TEST_HEADER") == "Test-Header"
+        assert from_wsgi_header("UNHANDLED_HEADER") is None
+        assert from_wsgi_header("CONTENT_TYPE") == "Content-Type"
+
+        # Check for the expected cache content
+        assert _cached_headers == {
+            "HTTP_TEST_HEADER": ("Test-Header", 1),
+            "UNHANDLED_HEADER": (None, 1),
+            "CONTENT_TYPE": ("Content-Type", 1),
+        }
+
+        # Check that the caching is not affecting the returned results
+        assert from_wsgi_header("HTTP_TEST_HEADER") == "Test-Header"
+        assert from_wsgi_header("UNHANDLED_HEADER") is None
+        assert from_wsgi_header("CONTENT_TYPE") == "Content-Type"
+
+        # Check that the cache is not growing indefinitely
+        _cached_headers.clear()
+
+        for i in range(_CACHE_HEADER_MAX_SIZE >> 1):
+            from_wsgi_header("HTTP_TEST_HEADER" + str(i))
+
+        for i in range(_CACHE_HEADER_MAX_SIZE):
+            from_wsgi_header("HTTP_TEST_HEADER" + str(i))
+
+        assert len(_cached_headers) == _CACHE_HEADER_MAX_SIZE
+
+        from_wsgi_header("HTTP_LAST_DROP")
+        assert len(_cached_headers) == (_CACHE_HEADER_MAX_SIZE >> 1) + 1
+        assert "HTTP_TEST_HEADER" + str(_CACHE_HEADER_MAX_SIZE - 1) not in _cached_headers
```

Comment:
Did you mean this here?

```suggestion
        assert "HTTP_TEST_HEADER" + str((_CACHE_HEADER_MAX_SIZE >> 1) - 1) not in _cached_headers
```

Also, worth checking the `HTTP_LAST_DROP` is in the cached headers.

Reply:
I think the original coding is correct; headers `HTTP_TEST_HEADER0...127` should have a count of 2, while the rest just a count of 1. Hence, after the `HTTP_LAST_DROP`, I expect headers with a count of 1 to be culled. I'll improve the test to also check for counts just to make sure :).